tests/adapters/test_protocol.py

```diff
@@ -22,10 +22,177 @@
+
+    def test_write_index2(self, written2):
+        assert written2._index == 1
+
+    def test_write_read_buffer1(self, written1):
+        assert written1._read_buffer == b"5"
+
+    def test_write_read_buffer2(self, written2):
+        assert written2._read_buffer == b"5"
+
+    def test_partial_write(self):
+        a = ProtocolAdapter([("written", 5)])
+        a.write_bytes(b"writ")
+        assert a._index == 0
+
+    def test_leftover_response(self):
+        a = ProtocolAdapter([("written", 5)])
+        a._read_buffer = b"5"
+        with raises(AssertionError):
+            a.write_bytes(b"written")
+
+    def test_no_response(self):
+        a = ProtocolAdapter([("written",)])
+        a.write_bytes(b"writ")
+        assert a._read_buffer == b""
+
+
+class Test_read:
+    @mark.parametrize("buffer, returned", ((b"\x03\x65", "\x03e"),
+                                           (b"Bytes", "Bytes"),
+                                           ))
+    def test_works(self, buffer, returned):
+        a = ProtocolAdapter()
+        a._read_buffer = buffer
+        assert a.read() == returned
+
+    def test_read_buffer_emptied(self):
+        a = ProtocolAdapter()
+        a._read_buffer = b"jklasdf"
+        a.read()
+        assert a._read_buffer == b""
+
+    def test_unsolicited_response(self):
+        a = ProtocolAdapter([(None, "Response")])
+        assert a.read() == "Response"
+
+    def test_not_enough_pairs(self):
+        a = ProtocolAdapter([("a",)])
+        a._index = 1
+        with raises(IndexError):
+            a.read()
+
+    def test_read_with_missing_write(self):
+        a = ProtocolAdapter([("a", "b")])
+        with raises(AssertionError):
+            a.read()
+
+
+class Test_read_bytes:
+    def test_read_full_buffer(self):
+        a = ProtocolAdapter()
+        a._read_buffer = b"Super 5"
+        assert a.read_bytes(3) == b"Sup"
+        assert a._read_buffer == b"er 5"
+
+    def test_read_empty_buffer(self):
+        a = ProtocolAdapter([(None, "Super 5")])
+        assert a.read_bytes(3) == b"Sup"
+        assert a._index == 1
+        assert a._read_buffer == b"er 5"
+
+    def test_read_without_write(self):
+        a = ProtocolAdapter([("written", 5)])
+        with raises(AssertionError):
+            a.read_bytes(3)
+
+    def test_no_messages(self):
+        a = ProtocolAdapter([])
+        with raises(IndexError):
+            a.read_bytes(3)
+
+
+def test_read_write_sequence():
+    """Test several consecutive writes and reads, including ask."""
+    a = ProtocolAdapter([("c1", "a1"), ("c2",), (None, "a3"), ("c4", "a4")])
+    a.write("c1")
+    assert a.read() == "a1"
+    a.write("c2")
+    assert a.read() == "a3"
+    assert a.ask("c4") == "a4"
+
+
+def test_write_and_read_with_and_without_bytes():
+    """Test writing and reading, normal and bytes in combination."""
+    a = ProtocolAdapter([("c1", "a1"), ("c4", "a4")])
+    a.write_bytes(b"c")
+    a.write("1")
```

Comment:
The second pair should be deleted, not part of this test.

I'm not sure if we should allow this write sequence. I think I can see a use case for piecewise binary writes (and probably reads). However, mixing binary and normal writes, and especially `write("1")` that _succeeds_ against an expected protocol of "c1" looks veryyy weird to me.

Reply:
See below about my communication thoughts.